Return focus after dialogs close

steverep opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
Adds an event handler to return focus, where possible, to the button or control that opened the dialog.  This is best practice for good accessibility.  Here are some details on the implementation:
- Finds and stores the active element and possibly dialog generically right before new dialog is shown
- Data is stored in the module in tag dictionary same as created elements, and typing was added for better code quality
- Handles the common special case where dialog is opened from a popup menu and returns focus to the trigger
- Handles the replace situation for more info / entity settings
- When dialog is opened from another dialog, focus method for `mwc-dialog` becomes the fallback in case of failure
### Limitations
1. Assumes the dialog fires the `dialog-closed` event.  The majority of the 70+ dialogs do, but about 20 or so need to be updated in future changes.
2. Assumes the element that opened the dialog is focusable, which is not always the case, but should be for keyboard accessibility (e.g. entity settings dialogs from configuration page and some more info dialogs from certain cards)

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: #10754 
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

steverep committed:
Add focus method for ha-icon-button

steverep committed:
Handle focus return after dialogs close

steverep commented:
Approvals is slower than usual... something I did? 😕 😄

zsarnett commented:
Bram is on Holiday 😄  I haven't been able to take a look yet.

steverep commented:
Ah I was wondering where he disappeared to.  Okay thanks for the heads up.

steverep commented:
Any chance this can get a review in time for April's beta?

bramkragten commented on src/dialogs/make-dialog-manager.ts:
```diff
@@ -60,11 +72,46 @@ export const showDialog = async (
+      LOADED[mainWindow.history.state.dialog].closedFocusDialog;
+  } else {
+    let focusedDialog: LoadedDialogInfo["closedFocusDialog"] = null;
+    let focusedElement = document.activeElement;
+    while (focusedElement?.shadowRoot?.activeElement) {
+      focusedElement = focusedElement.shadowRoot.activeElement;
```
Use https://github.com/home-assistant/frontend/blob/dev/src/common/dom/deep-active-element.ts#L1

bramkragten commented:
I wonder if we need this change, as `mwc-dialog` should handle this (it does here I think? https://material-components.github.io/material-web/demos/dialog/)

I think the issue is the location in the DOM where we add the dialogs probably? They are in a different place in the DOM then the element that triggers it...

Would have to look in to that

steverep commented:
There's no code in `mwc-dialog` to handle returning focus.  What OS/browser are you using on those demos?  Focus is not put back on the trigger for me in Chrome or Firefox on Windows.  It's just letting the browser handle it, which varies with browser.  For example, Firefox always throws focus back to the top of the document, whereas Chrome tends to throw focus further down the page.

You could make the argument that `mwc-dialog` should have a default to return to the last focused element, but that will only ever handle a subset of cases.  For example, `mwc-dialog` is never going to handle cases of replacing dialogs, trigger buttons becoming hidden, dialogs changing the DOM on close, etc.  That logic needs to be specific to the app.

bramkragten commented:
I'm using Chrome on MacOS

steverep commented:
> I'm using Chrome on MacOS

I don't have a Mac to reproduce that, but I'd be shocked if webkit was actively trying to return focus.  It doesn't do that on iOS.

In any case,leaving it up to the browser isn't a good or even consistent approach as I said, and it's definitely not `mwc-dialog` code.  All the component does is use [PolymerLabs/blocking-elements](https://github.com/PolymerLabs/blocking-elements) to set the `inert` and `aria-hidden="true"` attributes outside the dialog when it's open, and removes them when it's closed.  The only code regarding focus is to set it when the dialog is open, but only if the `dialogInitialFocus` attribute is set.

steverep commented:
@bramkragten  so what's the review verdict here?  Have I convinced you this code is needed independent of `mwc-dialog` in order to handle HA situations?

bramkragten commented:
Yeah, I'm seriously considering it, want to make it as easy as possible to maintain... Will get back to you this week.

steverep commented:
@balloob can you intercede here?  I've commented in the past that HA does not prioritize accessibility and this PR lingering on without a review is an example.  This implements simple focus management for dialogs - a fundamental piece of good keyboard accessibility.

zsarnett commented:
Please stop pinging developers. Bram advised he will be reviewing.

This is not about limiting Accessibility so don't try to pull on that string. This is about maintainability.

Normally your PRs go through without much question. This one adds a lot of code we have to maintain. Be patient

steverep commented:
> Please stop pinging developers. Bram advised he will be reviewing.

That was weeks ago and much work has come and merged since.  I don't see any problem with trying to call attention and start a discussion with the project's lead on prioritizing accessibility.  It was not meant to offend anyone.

> This is not about limiting Accessibility so don't try to pull on that string. This is about maintainability.

I didn't accuse the project of limiting accessibility.  I accused the project of not prioritizing it.  There's a huge difference.  If there is a way to make it more maintainable, then review and I'd be happy to implement it.  No one has done that.

> Normally your PRs go through without much question. This one adds a lot of code we have to maintain. Be patient

Please don't pretend like this is an exorbitant amount of code.  Again, many more lines have come and merged since.  And, just as food for thought, do you have any idea how often people with disabilities are told to be patient when trying to resolve accessibility or accommodation issues?  It's not about patience - it's just about treating those issues with the priority they deserve.

bramkragten commented on src/dialogs/make-dialog-manager.ts:
```diff
@@ -60,11 +72,46 @@ export const showDialog = async (
+      const slotRoot = focusedElement.assignedSlot?.getRootNode();
+      if (slotRoot instanceof ShadowRoot) {
+        const slotHost = slotRoot.host;
+        if (slotHost instanceof HaButtonMenu) {
+          // Use trigger button since popup menu will be hidden again
+          focusedElement = slotHost.querySelector('[slot="trigger"]');
+          break;
+        } else if (slotHost instanceof HaDialog) {
+          // Focus will return inside another dialog
+          focusedDialog = slotHost;
+        }
```
Can we instead add a function to the elements that need this, check if that function exists and then let that function return the correct element? So the logic is in the element and not in the dialog manager?

steverep replied:
My initial reaction was yes, but after thinking about how to code it, it won't really remove the logic from here because what to do with the returned value depends on the type of element.  This module would still need to know whether it's modifying the focus target or just recording an underlay dialog.

If you just want this logic out of the dialog module, I think I have a much better idea.  Give me a day or two to add a commit.

steverep replied:
I moved all of the active element/dialog searching to the `deepActiveElement` function, which now has 3 different modes.  So all of that is now in one place.  The rest of the code left in the dialog manager just grabs it, stores it, and sets the event handler.  See if that resolves your concerns.

Apologies for the delay... had to learn some unfamiliar Typescript then work around a known limitation.

steverep committed:
Expand deep active element utility to handle modes for dialog closing focus

steverep commented on src/common/dom/deep-active-element.ts:
```diff
@@ -1,8 +1,47 @@
+  modeType extends "normal" | "dialog" | "check"
+>(
+  root: DocumentOrShadowRoot = document,
+  mode?: modeType,
```
Making this change would decouple `mode` from the return type, which defeats the purpose of the conditional return.

Or are you just suggesting to remove `"normal"` as a possibility?

steverep commented on src/common/dom/deep-active-element.ts:
```diff
@@ -1,8 +1,47 @@
+import { HaDialog } from "../../components/ha-dialog";
+import { HaButtonMenu } from "../../components/ha-button-menu";
+
+export const deepActiveElement = <
```
I'm not following your suggestion.  Do you mean have the dialog manager call a new function that calls the original `deepActiveElement`?  It's impossible to use the original version because the checks need to be done at the document and every shadow root level.

steverep commented on src/common/dom/deep-active-element.ts:
```diff
@@ -1,8 +1,47 @@
+    activeElement = deeperElement ?? activeElement;
+  } while (deeperElement);
+
+  if (mode === "check") return false as any;
+  if (mode === "dialog") return [activeElement, activeDialog] as any;
+  return activeElement as any;
```
It needs to be there or Typescript throws an error.  This is the known limitation I referred to in my comment.  Apparently, conditional returns work great for the callers, but TS can't resolve it inside the function.  The work around is either to assert as any or use a bunch of overload signatures instead of the conditional.  See microsoft/TypeScript#33912 or any of the many duplicate issues.

steverep committed:
Add focus method to ha-button-menu and fix minor x/y type issue

steverep commented on src/common/dom/deep-active-element.ts:
```diff
@@ -1,8 +1,47 @@
+      const slotRoot = activeElement?.assignedSlot?.getRootNode();
+      if (slotRoot instanceof ShadowRoot) {
+        const slotHost = slotRoot.host;
+        if (slotHost instanceof HaButtonMenu) {
```
Working on a new more generic approach that will resolve this...

steverep committed:
Implement new strategy to return focus after dialogs close

The closed targets are now just a set based on a search for the active element and flagged ancestors, and no details of components are needed outside their constructor:
- Revert previous changes to deepActiveElement
- Add new utility to find flagged ancestors
- Flag certain components as safe focus targets
- Await next paint instead of just underlay dialog to be more robust

steverep commented:
@bramkragten  this is ready for another look.  I like this version better and I think it settles all your issues.

bramkragten commented on src/dialogs/make-dialog-manager.ts:
```diff
@@ -137,3 +168,31 @@ export const makeDialogManager = (
+    if (focusTarget instanceof HTMLElement) {
+      focusTarget.focus();
+      focusedElement = deepActiveElement();
+      if (focusedElement && focusedElement !== document.body) break;
```
We can return here right?

bramkragten commented:
Looks a lot better 👍

steverep committed:
Delete focus targets after use and return earlier from handler
